Lib/pathlib.py

```diff
@@ -1287,45 +1287,49 @@ def expanduser(self):
-        return self._walk(top_down, on_error, follow_symlinks)
-
-    def _walk(self, top_down, on_error, follow_symlinks):
-        # We may not have read permission for self, in which case we can't
-        # get a list of the files the directory contains. os.walk
-        # always suppressed the exception then, rather than blow up for a
-        # minor reason when (say) a thousand readable directories are still
-        # left to visit. That logic is copied here.
-        try:
-            scandir_it = self._scandir()
-        except OSError as error:
-            if on_error is not None:
-                on_error(error)
-            return
-
-        with scandir_it:
-            dirnames = []
-            filenames = []
-            for entry in scandir_it:
-                try:
-                    is_dir = entry.is_dir(follow_symlinks=follow_symlinks)
-                except OSError:
-                    # Carried over from os.path.isdir().
-                    is_dir = False
-
-                if is_dir:
-                    dirnames.append(entry.name)
-                else:
-                    filenames.append(entry.name)
-
-        if top_down:
-            yield self, dirnames, filenames
-
-        for dirname in dirnames:
-            dirpath = self._make_child_relpath(dirname)
-            yield from dirpath._walk(top_down, on_error, follow_symlinks)
+        stack = [(False, self)]
+
+        while stack:
+            must_yield, top = stack.pop()
+            if must_yield:
+                yield top
+                continue
+
+            # We may not have read permission for self, in which case we can't
+            # get a list of the files the directory contains. os.walk
+            # always suppressed the exception then, rather than blow up for a
+            # minor reason when (say) a thousand readable directories are still
+            # left to visit. That logic is copied here.
+            try:
+                scandir_it = top._scandir()
+            except OSError as error:
+                if on_error is not None:
+                    on_error(error)
+                continue
+
+            with scandir_it:
+                dirnames = []
+                filenames = []
+                for entry in scandir_it:
+                    try:
+                        is_dir = entry.is_dir(follow_symlinks=follow_symlinks)
+                    except OSError:
+                        # Carried over from os.path.isdir().
+                        is_dir = False
+
+                    if is_dir:
+                        dirnames.append(entry.name)
+                    else:
+                        filenames.append(entry.name)
+
+            if top_down:
+                yield top, dirnames, filenames
+            else:
+                stack.append((True, (top, dirnames, filenames)))
 
-        if not top_down:
-            yield self, dirnames, filenames
+            for dirname in reversed(dirnames):
```

Comment:
`os.scandir()` gives us names in an arbitrary order. Should we bother reversing them here?

Reply:
It seems to me that it's only arbitrary in the sense of it not having any clear rules about sorting. However, its order **is** stable which is why un-reversing them will make a few of the tests fail.

Reply:
It might be good to get a core dev opinion on this.

One thing that only just occurred to me: users of `top_down=True` might expect us to visit directories in the same order they appear in the yielded `dirnames`. Or maybe not, as it's not documented anywhere.

Reply:
1. I believe that users will expect Path.walk ordering to be consistent with the ordering of the recursive `os.walk` that existed for 20+ years. And I think our tests verify that it is consistent.
2. We only mention scandir in the error handling section in `walk`'s docs so I think the absolute majority of users do not even think about scandir being used within walk, and definitely do not know much about its ordering rules.

Reply:
@AlexWaygood what do you think? The `reversed()` call is there to ensure we visit subdirectories in the same order as before this patch, but we don't document that order, and in fact it's arbitrary (but stable across runs) as we're using `os.scandir()` under-the-hood. Should it be removed for simplicity's sake? (+ tiny performance improvement, I assume)

Reply:
If we decide to remove the `reversed()` call then the tests will need adjustment. If possible I'd love for them to stop making assertions about the order of items in `dirnames` and `filenames`, and stop making assertions about the order in which sibling directories are visited. Is that doable @Ovsyanka83, or is it a bit of a horror show?

Reply:
I do not think like it is a big problem. Many tests will need to be reworked and some thought will need to be put into making sure our tests stay effective but it is definitely doable.

Reply:
I don't _think_ the `reversed()` call should have a noticeable impact on performance, due to the fact that `reversed()`, unlike `sorted()`, is a _lazy_ iterator: it uses a specialised iterator to iterate over lists backwards, rather than creating a new list in memory:

```pycon
>>> reversed([])
<list_reverseiterator object at 0x00000299ED6A2CB0>
```

(Having said that, if you've got numbers to prove me wrong, I'd be very interested to see them ;)

I also agree with @Ovsyanka83 that, while `Path.walk()` is a new function (so we're not bound by backwards-compatibility constraints), it certainly seems like it would probably be the least surprising thing to match the iteration order that users have grown accustomed to with `os.walk()`.

Reply:
Makes sense to me, thank you both.